builder/optimizations.cc

```diff
@@ -472,27 +448,41 @@ std::tuple<stmt, stmt, stmt> split_body(const stmt& body, span<const symbol_id>
   stmt new_body_after;
```

Comment:
I think this implementation is O(n^2) (each stmt makes a new vector and copies all previous stmts). It's basically as if `std::vector` grew by one at a time instead of a factor.

I think this should have `std::vector<stmt> before, new_body, after` and just directly find the begin/end of each chunk, something like:

```
if (const block* b = body.as<block>()) {
  auto depends_on_i = [&](std::vector<stmt>::iterator i) { return depends_on(*i, vars); };
  auto end_before = std::find_if(b->stmts.begin(), b->stmts.end(), depends_on_i));
  before = {b->stmts.begin(), end_before)};
  if (end_before != b->stmts.end()) {
    // We know that this is not b->stmts.rend() because otherwise the entire body would be in `before` above.
    // TODO: Not sure if this should be + or - 1.
    auto end_body = std::find_if(b->stmts.rbegin(), b->stmts.rend(), depends_on_i).base() - 1;
    new_body = {end_before, end_body)};
    after = {end_body, b->stmts.end()};
  }
} else if (depends_on(body, vars)) {
  new_body.push_back(body);
} else {
  before.push_back(body);
}
```

This probably has bugs, but I think the idea is sound and just needs a little tweaking hopefully.

Reply:
Fair, was just doing the simple adaptation of the old code, stand by